### ISSUE TITLE
grub: error: symbol grub_calloc not found (after upgrade to 21.05)

### ISSUE THREAD

**Issue by ppenguin** [NixOS/nixpkgs#127925]
**Describe the bug**
Upgrade to 21.05 renders non-UEFI system unbootable: `grub error: symbol grub_calloc not found`

**To Reproduce**
1. update to 21.05 on a legacy BIOS machine (i.e. no UEFI)
2. observe you cannot boot anymore: `grub error`: `error: symbol grub_calloc not found`

**Expected behavior**

System boots after upgrade.

**Screenshots**

(screenshot not my own, but identical): https://i.stack.imgur.com/Y4dXN.jpg

**Additional context**

Could it be an issue with the specific `grub-2.06-rc1` version?

**Notify maintainers**

**Metadata**
(after reverting back to `20.09` and being able to boot again:)
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.122, NixOS, 20.09.4346.d6a5be86b6a (Nightingale)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.11`
 - channels(root): `"nixos-20.09.4346.d6a5be86b6a, nixos-unstable-21.11pre297098.7e567a3d092"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by srhb** [comment]
@ppenguin We had this on one of our servers, where I think we determined that it was caused by an incompatible (old) version of grub in firmware being unable to load the _newer_ version. Are you sure you don't have an old version being loaded from somewhere? I realize you're not on UEFI, but I imagine it could come from all manner of places if you haven't explicitly reinstalled it.

TL;DR: I'm not (yet) convinced there's an actual problem with this version of grub, except for a backwards incompatibility that's really not advertised in a nice way.

**Comment by ppenguin** [comment]
@srhb thanks, that might play a role indeed, I remember on the first install (fedora a long time ago) the selection of the boot device was a bit involved due to the (unused) RAID functionality (it's a HP microserver gen8) and the desire to boot from a non-data drive (i.e. not connected to the backplane).

Would you know of a good way how/where to find potentially polluting old `grub` remnants and how to get rid of them?

**Comment by srhb** [comment]
Sorry, no. My first bet if you're sure there's no EFI involved is to `nixos-rebuild boot --install-bootloader` -- but that'll only work for whatever your system config currently tries to install, not any remnants that might be lying around not reflected by your system config.

**Comment by wamserma** [comment]
Running an older desktop with MBR here and I did not encounter the bug when upgrading.
Have you tried listing all devices in [`boot.loader.grub.devices`](https://search.nixos.org/options?channel=21.05&show=boot.loader.grub.devices&from=0&size=50&sort=relevance&query=grub) to overwrite any possible remnants?

**Comment by allgreed** [comment]
A workaround (requires a live installation media and internet connection, obviously worked for me):
```bash
# from live installation, mount your / to /mnt
nix-shell -p "(import (builtins.fetchurl https://matthewbauer.us/generate-versions.nix) {}).grub2.\"2.04\""
# repeat per boot drive
grub-install --root-directory=/mnt [your drive]
```

Related:
- literally any ask-ubuntu / stack page telling to reinstall grub
- http://matthewbauer.us/blog/all-the-versions.html

I think it might be possible to substitute the realGrub package used by [nixos grub module](https://github.com/NixOS/nixpkgs/blob/nixos-21.05/nixos/modules/system/boot/loader/grub/grub.nix) so that subsequent application of `nixos-rebuild` don't install the latest grub therefore breaking the boot again, however I haven't unlocked ancient secrets™ yet, so no idea how to do it.

Note to maintainers: can a sensible check be implemented to prevent this kind of issues (new grub breaking BIOS boot) in the future? I'll take a reproducible non-working printer any time over a working statefull mess, however booting is kind of critical. Again, I've got no idea how to achieve this, but I'm willing to try to do it / help / make a damn sandwich for someone working on it should an idea emerge.

**Comment by ppenguin** [comment]
I was feeling bold and tried again on my HP Microserver Gen8 to go directly to 21.11 this time, and got the same issue.
Needless to say it's a pain since I can only do recovery stuff via `iLO` and a remote mounted ISO, which is not very robust (it often unmounts during boot).

Thanks to @allgreed I got grub going again with the above recipe, but somehow the kernel wasn't found by it.

Adding one more grub device (my MBRs are a bit funky including an SD card due to some peculiar controller stuff if you don't want boot from RAID) after booting from install ISO (like @wamserma suggested) and just running `nixos-install` over the old installation appears to have solved the issue, i.e. it now boots correctly with `grub-2.06`

**Comment by ppenguin** [comment]
Closing this to cleanup, I haven't tested whether this still occurs for a long time
